crates/schemalate/src/firebolt/firebolt_projections.rs

```diff
@@ -0,0 +1,499 @@
+use std::collections::{BTreeMap, HashSet};
+
+use protocol::flow::{inference::Exists, materialization_spec::Binding, CollectionSpec};
+use protocol::materialize::{constraint, Constraint};
+
+/*pub fn validate_selected_fields(
+    constraints: BTreeMap<String, Constraint>,
+    proposed: Binding,
+) -> Result<(), InvalidSelectedFields> {
+}*/
+
+pub fn validate_new_projection(proposed: Binding) -> BTreeMap<String, Constraint> {
+    proposed
+        .collection
+        .unwrap()
+        .projections
+        .iter()
+        .map(|projection| {
+            let constraint = {
+                if projection.is_primary_key {
+                    Constraint {
+                        r#type: constraint::Type::LocationRequired.into(),
+                        reason: "All locations that are part of the collection key are required."
+                            .to_string(),
+                    }
+                } else if projection.ptr.len() == 0 {
+                    // root document
+                    Constraint {
+                        r#type: constraint::Type::LocationRecommended.into(),
+                        reason: "The root document should usually be materialized.".to_string(),
+                    }
+                } else if let Some(infer) = &projection.inference {
+                    if infer.types.len() != 1 {
```

Comment:
`null` is one of the possible types; it should be valid to materialize a nullable integer, for example, right?

Consider using `json::schema::types::Set::from_iter()`, followed by `Set::is_single_scalar_type`  and/or composed type set literals.